File: internal/validation/registration.go
package validation

import (
	"errors"
	"regexp"
	"social/internal/models"
	"strings"
	"time"
)

func ValidateRegisterData(userData *models.UserRegistration) error {
	NormalizeRegisterData(userData)

	if err := validateNames(&userData.FirstName); err != nil {
		return err
	}
	if err := validateNames(&userData.LastName); err != nil {
		return err
	}

	if err := validateUserName(&userData.UserName); err != nil {
		return err
	}

	if err := validateEmail(&userData.Email); err != nil {
		return err
	}

	if err := validateDOB(&userData.DOB); err != nil {
		return err
	}

	if err := validatePassword(userData.Password); err != nil {
		return err
	}

	if err := validateAbout(&userData.About); err != nil {
		return nil
	}

	return nil
}

// take the data and normilize it for the database and the validation as removing trail and leading spaces and capitalize or lowerize the chracters
func NormalizeRegisterData(userData *models.UserRegistration) {
	// trim spaces
	userData.FirstName = strings.Trim(userData.FirstName, " ")
	userData.LastName = strings.Trim(userData.LastName, " ")
	userData.UserName = strings.Trim(userData.UserName, " ")
	userData.Email = strings.Trim(userData.Email, " ")
	userData.Password = strings.Trim(userData.Password, " ")
	userData.About = strings.Trim(userData.About, " ")

	// normalize
	if userData.FirstName != "" && len(userData.FirstName) >= 2 {
		userData.FirstName = strings.ToUpper(userData.FirstName[:1]) + userData.FirstName[1:]
	}
	if userData.LastName != "" && len(userData.LastName) >= 2 {
		userData.LastName = strings.ToUpper(userData.LastName[:1]) + userData.LastName[1:]
	}
	if userData.UserName != "" {
		userData.UserName = strings.ToLower(userData.UserName)
	}
	if userData.Email != "" {
		userData.Email = strings.ToLower(userData.Email)
	}
	

}

func validateNames(name *string) error {
	if len(*name) < 2 || len(*name) > 15 {
		return errors.New("Error: first/last name must be between 3 and 15 characters")
	}

	nameReg, err := regexp.Compile(`^[A-Z][a-zA-Z]*$`)
	if err != nil {
		return err
	}

	if !nameReg.MatchString(*name) {
		return errors.New("Error: name format is incorrect")
	}

	return nil
}

func validateUserName(username *string) error {
	if len(*username) == 0 {
		return nil
	}
	if len(*username) < 3 || len(*username) > 12 {
		return errors.New("Error: usernam length must be between 3 and 12 characters")
	}

	usernameReq, err := regexp.Compile("^[a-z0-9_-]+$")
	if err != nil {
		return err
	}

	if !usernameReq.MatchString(*username) {
		return errors.New("Error: invalid username format")
	}

	return nil
}

func validateEmail(email *string) error {
	if len(*email) < 5 || len(*email) > 75 {
		return errors.New("Error: email length must be between 5 and 75 characters")
	}

	emailReg, err := regexp.Compile(`^[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}$`)
	if err != nil {
		return err
	}

	if !emailReg.MatchString(*email) {
		return errors.New("Error: invalid email")
	}

	return nil
}

func validateDOB(dob *time.Time) error {
	now := time.Now()

	if dob.After(now) {
		return errors.New("date of birth is invalid")
	}

	tooOld := now.AddDate(-120, 0, 0)

	if dob.Before(tooOld) {
		return errors.New("date of birth is invalid")
	}

	tooYoung := now.AddDate(-12, 0, 0)
	if dob.After(tooYoung) {
		return errors.New("date of birth is invalid")
	}

	return nil
}

func validatePassword(pass string) error {
	if len(pass) < 8 {
		return errors.New("invalid password: password must be at least 8 characters long")
	}

	if len(pass) > 75 {
		return errors.New("invalid password: password must be less than 75 characters long")
	}

	numberCheck, err := regexp.MatchString(`[0-9]`, pass)
	if err != nil {
		return errors.New("invalid password: password format is wrong")
	}

	if !numberCheck {
		return errors.New("invalid password: password must contain a number")
	}

	specialCharacterCheck, err := regexp.MatchString(`[!@#$%*&^?]`, pass)
	if err != nil {
		return errors.New("invalid password: password format is wrong")
	}

	if !specialCharacterCheck {
		return errors.New("invalid password: password must contain a special character: ! @ # $ % * & ^ ?")
	}

	alphaCheck, err := regexp.MatchString(`[a-zA-Z]`, pass)
	if err != nil {
		return errors.New("invalid password: password format is wrong")
	}

	if !alphaCheck {
		return errors.New("invalid password: password must contain alphabetic characters")
	}

	forbiddenCheck, err := regexp.MatchString(`[^a-zA-Z0-9!@#$%*&^?]`, pass)
	if err != nil {
		return errors.New("invalid password: password format is wrong")
	}

	if forbiddenCheck {
		return errors.New("invalid password: password contains forbidden characters")
	}

	return nil
}

func validateAbout(about *string) error {
	if len(*about) == 0 {
		return nil
	}
	if len(*about) > 1000 {
		return errors.New("Error: about is too long")
	}
	return nil
}

File: frontend/src/src/helpers/validators/registration.js
export function validateName(name = "") {
    if (name.length < 3 || name.length > 15) {
        return "Name must be between 2 and 15 characters"
    }

    const reg = /^[A-Z][a-zA-Z]*$/

    if (!reg.test(name)) {
        return "Name format is incorrect"
    }

    return ""
}

export function validateUsername(username = "") {
    if (username.length === 0) {
        return ""
    }

    if (username.length < 3 || username.length > 12) {
        return "Username must be between 3 and 12 characters"
    }

    const reg = /^[a-z0-9_-]+$/

    if (!reg.test(username)) {
        return "Username can only contain a-z, 0-9, '-' and '_'"
    }

    return ""
}

export function validateEmail(email = "") {
    if (email.length < 5 || email.length > 75) {
        return "Email must be between 5 and 75 characters"
    }

    const reg = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/

    if (!reg.test(email)) {
        return "Invalid email"
    }

    return ""
}

export function validatePassword(password = "") {
    if (password.length < 8) {
        return "Password must be at least 8 characters long"
    }

    if (password.length > 75) {
        return "Password must be less than 75 characters long"
    }

    const numberReg = /[0-9]/
    const specialReg = /[!@#$%*&^?]/
    const alphaReg = /[a-zA-Z]/
    const forbiddenReg = /[^a-zA-Z0-9!@#$%*&^?]/

    if (!numberReg.test(password)) {
        return "Password must contain a number"
    }

    if (!specialReg.test(password)) {
        return "Password must contain a special character: ! @ # $ % * & ^ ?"
    }

    if (!alphaReg.test(password)) {
        return "Password must contain alphabetic characters"
    }

    if (forbiddenReg.test(password)) {
        return "Password contains forbidden characters"
    }

    return ""
}

export function validateAbout(about = "") {
    if (about.length === 0) {
        return ""
    }

    if (about.length > 1000) {
        return "About is too long"
    }

    return ""
}

export function validateDOB(dob = "") {
    if (!dob) {
        return "Date of birth is invalid"
    }

    const date = new Date(dob)
    const now = new Date()

    const tooOld = new Date()
    tooOld.setFullYear(now.getFullYear() - 120)

    const tooYoung = new Date()
    tooYoung.setFullYear(now.getFullYear() - 12)

    if (date > now || date < tooOld || date > tooYoung) {
        return "Date of birth is invalid"
    }

    return ""
}

export function handleNameInput(value = "") {
    let result = value.replace(/\s+/g, " ").trimStart()

    if (result.length > 0) {
        result = result.charAt(0).toUpperCase() + result.slice(1).toLowerCase()
    }

    return result
}

export function handleUsernameInput(value = "") {
    return value.toLowerCase().trim()
}

export function handleEmailInput(value = "") {
    return value.toLowerCase().trim()
}

export function validateAvatar(file) {
    if (!file || file.size === 0) {
        return ""
    }

    const allowedTypes = ["image/jpeg", "image/png", "image/gif"]

    if (!allowedTypes.includes(file.type)) {
        return "Avatar must be a JPG, PNG, or GIF file"
    }

    return ""
}